fri/src/two_adic_pcs.rs

```diff
@@ -0,0 +1,539 @@
+            &fri_challenges,
+            &reduced_openings,
+        )
+        .map_err(VerificationError::FriError)?;
+
+        Ok(())
+    }
+}
+
+#[instrument(skip_all)]
+fn compute_inverse_denominators<F: TwoAdicField, EF: ExtensionField<F>, M: Matrix<F>>(
+    mats_and_points: &[(Vec<M>, &[Vec<EF>])],
+    coset_shift: F,
+) -> LinearMap<EF, Vec<EF>> {
+    let mut max_log_height_for_point: LinearMap<EF, usize> = LinearMap::new();
+    for (mats, points) in mats_and_points {
+        for (mat, points_for_mat) in izip!(mats, *points) {
+            let log_height = log2_strict_usize(mat.height());
+            for &z in points_for_mat {
+                if let Some(lh) = max_log_height_for_point.get_mut(&z) {
+                    *lh = core::cmp::max(*lh, log_height);
+                } else {
+                    max_log_height_for_point.insert(z, log_height);
+                }
+            }
+        }
+    }
+    let max_log_height = *max_log_height_for_point.values().max().unwrap();
+    // Compute the largest subgroup we will use, in bitrev order.
+    let mut subgroup = cyclic_subgroup_coset_known_order(
+        F::two_adic_generator(max_log_height),
+        coset_shift,
+        1 << max_log_height,
+    )
+    .collect_vec();
+    reverse_slice_index_bits(&mut subgroup);
+    max_log_height_for_point
+        .into_iter()
+        .map(|(z, log_height)| {
+            (
+                z,
+                batch_multiplicative_inverse(
+                    &subgroup[..(1 << log_height)]
+                        .iter()
+                        .map(|&x| EF::from_base(x) - z)
+                        .collect_vec(),
+                ),
+            )
+        })
+        .collect()
+}
+
+struct PowersReducer<F: Field, EF> {
+    powers: Vec<EF>,
+    // If EF::D = 2 and powers is [01 23 45 67],
+    // this holds [[02 46] [13 57]]
+    transposed_packed: Vec<Vec<F::Packing>>,
+}
+
+impl<F: Field, EF: ExtensionField<F>> PowersReducer<F, EF> {
+    fn new(base: EF, max_width: usize) -> Self {
+        let powers: Vec<EF> = base
+            .powers()
+            .take(max_width.next_multiple_of(F::Packing::WIDTH))
+            .collect();
+
+        let transposed_packed: Vec<Vec<F::Packing>> = (0..EF::D)
+            .map(|i| {
+                F::Packing::pack_slice(&powers.iter().map(|a| a.as_base_slice()[i]).collect_vec())
+                    .to_vec()
+            })
+            .collect();
+
+        Self {
+            powers,
+            transposed_packed,
+        }
+    }
+
+    // Compute sum_i base^i * x_i
+    fn reduce_ext(&self, xs: &[EF]) -> EF {
+        self.powers.iter().zip(xs).map(|(&pow, &x)| pow * x).sum()
+    }
+
+    // Same as `self.powers.iter().zip(xs).map(|(&pow, &x)| pow * x).sum()`
+    // For some reason, this is 2x faster than naive in benches but has
+    // the exact same performance in the actual implementation (for BabyBear).
+    // More investigation is needed.
```

Comment:
No clue why this happens. In microbenches, packing is clearly 2-3x faster, but in the actual keccak PCS, `reduce rows [ 237ms | 6.73% ]` stays the same whether it's packed or the naive `sum()`. I checked that the 237 ms are actually spent reducing, and a few different permutations of the code but couldn't get it to speed up. ¯\(ツ)/¯

Reply:
FWIW I get a pretty close result
```
sample 4: slow: 381.161291ms fast: 376.824708ms
```
were you using a release build?

Reply:
Interestingly in the microbench, `reduce_base` (the vectorized one) seems to become like 2x faster (relative to input size) on smaller inputs, like 99 `xs` instead of 99.

Reply:
Hm, good catch. Not sure how to record memory performance in Instruments, but cpu must not be the bottleneck

Reply:
Weird. No dice trying to speed this up, there are more stalls in `fast` than `slow` but I'm not sure where exactly they're coming from. There's a decent amount of latency in the reads, but they should be prefetched well since it's linearly iterating? Idk, would need to start messing with prefetching cache lines or something

![image](https://github.com/Plonky3/Plonky3/assets/7348605/485c2e10-e200-4ef1-b68b-2698e9e0a670)

Reply:
I mean, it's probably because `slow` only does one iteration over `self.powers`, while `fast` does `D` iterations for extension degree `D`. So it's probably evicting `xs` from L1 at some point, even though it should keep it around for the next iter

Reply:
Nvm got it fast by using my old vertical_packing thing.